In [2]:
import numpy as np
import pandas as pd

In [3]:
#FoodBalanceSheets_E_Africa_NOFLAG
df=pd.read_csv("FoodBalanceSheets_E_Africa_NOFLAG.csv", encoding="latin-1")

In [4]:
# show the head of the dataset
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

Question 1
What is the total number of unique countries in the dataset?


In [ ]:
df['Area'].nunique()

Question 3
Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6


In [ ]:
my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

It gives a Type Error

Question 4
Answer the following questions based on the African food production dataset provided by the FAO website already provided

What is the total sum of Wine produced in 2015 and 2018 respectively?

In [5]:
grouped_winedata = df.groupby('Item')['Y2015','Y2018'].sum()

print(grouped_winedata)

                         Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]


C:\Users\adeba\AppData\Local\Temp\ipykernel_10360\906988447.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  grouped_winedata = df.groupby('Item')['Y2015','Y2018'].sum()


The value of wine is 4251.81 and 4039.32 respectively

Question 5
Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

In [7]:
Area_2017 = df[['Y2017', 'Area']]
Area2017_data = Area_2017.groupby('Area').sum()
 
Area2017_data.sort_values(by='Y2017', ascending=True)



Y2017
Area                                   
Sudan (former)                     0.00
Ethiopia PDR                       0.00
Comoros                           59.84
Seychelles                       442.34
Sao Tome and Principe          12662.63
Cabo Verde                     14650.74
Guinea-Bissau                  19102.77
Lesotho                        21267.96
Botswana                       22101.30
Djibouti                       22729.91
Gambia                         23154.18
Gabon                          27979.64
Liberia                        29342.20
Namibia                        29874.89
Central African Republic       29937.00
Congo                          41181.68
Togo                           49841.88
Mauritius                      51114.83
Eswatini                       54343.33
Sierra Leone                   55311.33
Chad                           71594.68
Rwanda                         73663.69
Zimbabwe                       75919.34
Senegal                        95681.15
Guinea                         98138.87
Burkina Faso                  101855.07
Zambia                        103223.77
Tunisia                       124167.20
Benin                         124771.22
Niger                         126707.58
Madagascar                    131197.73
Mali                          149928.33
Mauritania                    156665.46
Mozambique                    161407.98
Malawi                        181098.71
Uganda                        213950.38
Côte d'Ivoire                 224599.01
Angola                        229159.57
Cameroon                      232030.43
Sudan                         239931.92
Kenya                         264660.66
United Republic of Tanzania   322616.85
Algeria                       325644.27
Ghana                         337599.06
Morocco                       388495.36
Ethiopia                      448683.76
South Africa                  517590.54
Egypt                         866379.92
Nigeria                      1483268.23

Guinea Bissau is the seventh lowest sum in 2017


How would you check for the number of rows and columns in a pandas DataFrame named df?

In [8]:
df.shape

(60943, 12)

What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

In [11]:
# mean of the data in 2017
mean_2017= round(df['Y2017'].mean(),2)
#standard deviation of the data in 2017
std_2017 = round(df['Y2017'].std(),2)
print ('The mean is ',mean_2017,'The standard deviation is ',std_2017)

The mean is  140.92 The standard deviation 1671.86


What is the total Protein supply quantity in Madagascar in 2015?

In [13]:
# show protein supply in madagascar 2015 and sum them up
dfarea= df[df['Area']=='Madagascar'].copy()
dfprotein = dfarea[dfarea['Element']=='Protein supply quantity (g/capita/day)']
dfprotein['Y2015'].sum()


173.04999999999998

Question 3

the DataFrame has 100 rows (observations) and 3 columns (features), so there are 3 features and 100 observations.

Question 6
If you have the following list

lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = [‘Age’,’Nationality’,’Overall’]

In [19]:
lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = ['Age','Nationality','Overall']
pd.DataFrame(lst, columns=col, index=[1,2,3])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

The Answer is 4251.81 and 4039.32 respectively

Which year had the least correlation with ‘Element Code’?

In [18]:
correlation = df[['Y2014', 'Y2015', 'Y2016','Y2017']].corrwith(df['Element Code'])

print(correlation)

Y2014    0.024457
Y2015    0.023889
Y2016    0.023444
Y2017    0.024254
dtype: float64


Therefore the year with the least correlation is Year 2016


Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation

In [19]:
grouped_data12 = df.groupby('Element')['Y2014','Y2015','Y2016','Y2017'].sum()

print(grouped_data12)

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00  

C:\Users\adeba\AppData\Local\Temp\ipykernel_10360\2917983886.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  grouped_data12 = df.groupby('Element')['Y2014','Y2015','Y2016','Y2017'].sum()


What is the total number and percentage of missing data in 2014 to 3 decimal places?

In [ ]:
# Number of missing data in 2014
y2014_missing= df['Y2014'].isnull().sum()
print(y2014_missing)
# percent missing
y2014_notmissing = df['Y2014'].notnull().sum()
y2014_percentmissing =(100*y2014_missing/(y2014_missing+y2014_notmissing))
print(round(y2014_percentmissing,3))

Therefore the answer is 1589 and 2.607%

Consider the following list of tuples:

y = [(2, 4), (7, 8), (1, 5, 9)]

How would you assign element 8 from the list to a variable x?

In [15]:
y = [(2, 4), (7, 8), (1, 5, 9)]
x=y[1][-1]
print(x)

8


In [ ]:
x=y[1][-1] x=y[1][1]


A pandas Dataframe with dimensions (100,3) has how many features and observations?

What would be the output for?

S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1

In [22]:
S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

'e'

Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

In [21]:
Area_2017 = df[['Y2017', 'Area']]
Area2017_data = Area_2017.groupby('Area').sum()
 
Area2017_data.sort_values(by='Y2017', ascending=False)



Y2017
Area                                   
Nigeria                      1483268.23
Egypt                         866379.92
South Africa                  517590.54
Ethiopia                      448683.76
Morocco                       388495.36
Ghana                         337599.06
Algeria                       325644.27
United Republic of Tanzania   322616.85
Kenya                         264660.66
Sudan                         239931.92
Cameroon                      232030.43
Angola                        229159.57
Côte d'Ivoire                 224599.01
Uganda                        213950.38
Malawi                        181098.71
Mozambique                    161407.98
Mauritania                    156665.46
Mali                          149928.33
Madagascar                    131197.73
Niger                         126707.58
Benin                         124771.22
Tunisia                       124167.20
Zambia                        103223.77
Burkina Faso                  101855.07
Guinea                         98138.87
Senegal                        95681.15
Zimbabwe                       75919.34
Rwanda                         73663.69
Chad                           71594.68
Sierra Leone                   55311.33
Eswatini                       54343.33
Mauritius                      51114.83
Togo                           49841.88
Congo                          41181.68
Central African Republic       29937.00
Namibia                        29874.89
Liberia                        29342.20
Gabon                          27979.64
Gambia                         23154.18
Djibouti                       22729.91
Botswana                       22101.30
Lesotho                        21267.96
Guinea-Bissau                  19102.77
Cabo Verde                     14650.74
Sao Tome and Principe          12662.63
Seychelles                       442.34
Comoros                           59.84
Sudan (former)                     0.00
Ethiopia PDR                       0.00

Nigeria had the highest

Which of the following dataframe methods can be used to access elements across rows and columns?

In [20]:
df.iloc[:]

Area Code      Area  Item Code           Item  Element Code  \
0              4   Algeria       2501     Population           511   
1              4   Algeria       2501     Population          5301   
2              4   Algeria       2901    Grand Total           664   
3              4   Algeria       2901    Grand Total           674   
4              4   Algeria       2901    Grand Total           684   
...          ...       ...        ...            ...           ...   
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit     Y2014  \
0               Total Population - Both sexes     1000 persons  38924.00   
1                    Domestic supply quantity      1000 tonnes      0.00   
2               Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3      Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4          Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                       ...              ...       ...   
60938                                    Food      1000 tonnes     42.00   
60939     Food supply quantity (kg/capita/yr)               kg      3.06   
60940           Food supply (kcal/capita/day)  kcal/capita/day      3.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day      0.10   
60942      Fat supply quantity (g/capita/day)     g/capita/day      0.04   

          Y2015     Y2016     Y2017     Y2018  
0      39728.00  40551.00  41389.00  42228.00  
1          0.00      0.00      0.00      0.00  
2       3379.00   3372.00   3341.00   3322.00  
3         94.35     94.72     92.82     91.83  
4         79.36     77.40     80.19     77.28  
...         ...       ...       ...       ...  
60938     46.00     33.00     19.00     16.00  
60939      3.33      2.35      1.33      1.08  
60940      4.00      3.00      1.00      1.00  
60941      0.11      0.08      0.04      0.04  
60942      0.05      0.03      0.02      0.01  

[60943 rows x 12 columns]

Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017

In [17]:
df.groupby('Element')['Y2017'].sum()

Element
Domestic supply quantity                  2088198.10
Export Quantity                            182338.80
Fat supply quantity (g/capita/day)          10253.84
Feed                                       223705.68
Food                                      1258888.28
Food supply (kcal/capita/day)              454681.00
Food supply quantity (kg/capita/yr)         48690.04
Import Quantity                            294559.09
Losses                                     160614.00
Other uses (non-food)                       91645.97
Processing                                 292836.00
Production                                2030056.89
Protein supply quantity (g/capita/day)      11842.45
Residuals                                   35500.00
Seed                                        24870.14
Stock Variation                             54316.91
Total Population - Both sexes             1112641.00
Tourist consumption                            91.00
Name: Y2017, dtype: float64

Question 10

Therefore the year with the least correlation is Year 2016

Question 12

In [63]:
import warnings

# Suppress all warnings
warnings.filterwarnings('ignore')

grouped_data12 = df.groupby('Element')['Y2014','Y2015','Y2016','Y2017'].sum()

print(grouped_data12)

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00  

2014 has the highest sum of stock variation